File: server/algorithm.py
import math
import numpy as np


class Algorithm:
    def __init__(self, config, params):
        self.params = params
        self.config = config
        self.rng = np.random.default_rng()

        self.force = np.ones(1)

        self.hue_vel = np.zeros(1)
        self.linear_sat_vel = np.zeros(1)
        self.linear_val_vel = np.zeros(1)

        self.linear_sat = np.zeros(1)
        self.linear_val = np.zeros(1)

        self.hue = self.rng.random(1)
        self.sat = np.zeros(1)
        self.val = np.zeros(1)

        self.hue_snake = np.full(self.config.num_pixels, self.hue[0])
        self.sat_snake = np.zeros(self.config.num_pixels)
        self.val_snake = np.zeros(self.config.num_pixels)

    def set_params(self, params):
        self.params = params
        self.params.s = math.pow(self.params.s, 1.0 / self.config.scaling_root)
        self.params.v = math.pow(self.params.v, 1.0 / self.config.scaling_root)

    def update(self):
        for position, velocity, target, tolerance, wraps in (
            (self.hue, self.hue_vel, self.params.h, self.params.dh, True),
            (self.linear_sat, self.linear_sat_vel, self.params.s, self.params.ds, False),
            (self.linear_val, self.linear_val_vel, self.params.v, self.params.dv, False),
        ):
            # Sometimes the fastest way to get to the target is to decriment
            # the hue, and wrap around from 0 -> 1. The "inverse target" is
            # the opposite point on the hue circle. By adding or subtracting
            # 1, we make the comparisons below simpler. Because we wrap the
            # values at the end of this function, the offset here doesn't
            # effect the final value.
            inverse_target = (target + 0.5) % 1
            if wraps:
                if inverse_target > target:
                    np.subtract(position, 1.0, where=position > inverse_target, out=position)
                else:
                    np.add(position, 1.0, where=position < inverse_target, out=position)

            # Calculate random forces on the pixel due to temperature
            self.rng.random(1, out=self.force)
            np.subtract(self.force, self.params.t, out=self.force)
            np.clip(self.force, None, 0.0, out=self.force)
            np.add(self.force, self.params.t / 2.0, where=self.force != 0, out=self.force)
            np.sign(self.force, out=self.force)
            np.multiply(self.force, self.config.delta, out=self.force)

            # Apply restoring forces
            restoring_delta = self.config.delta * self.config.restoring_force
            if wraps or target == 0.0:
                np.subtract(self.force, restoring_delta, where=position > target + tolerance, out=self.force)

            np.add(self.force, restoring_delta, where=position < target - tolerance, out=self.force)

            # Acceleration = Force / Mass
            np.divide(self.force, self.config.mass, out=self.force)

            # Velocity += Acceleration
            np.add(velocity, self.force, out=velocity)

            # Position += Velocity
            np.add(position, velocity, out=position)

            # Velocity *= friction
            np.multiply(velocity, self.config.friction, out=velocity)

            if wraps:
                np.mod(position, 1.0, out=position)
            else:
                np.clip(position, 0.0, 1.0, out=position)

        np.power(self.linear_sat, 1.0 / self.config.scaling_root, out=self.sat)
        np.power(self.linear_val, 1.0 / self.config.scaling_root, out=self.val)

        self.hue_snake = np.roll(self.hue_snake, 1)
        self.sat_snake = np.roll(self.sat_snake, 1)
        self.val_snake = np.roll(self.val_snake, 1)
        self.hue_snake[0] = self.hue[0]
        self.sat_snake[0] = self.sat[0]
        self.val_snake[0] = self.val[0]

File: src/algorithm.c
#include "algorithm.h"

#include <stdint.h>
#include <zephyr/random/random.h>

#include "params.h"

void algorithm_init(Algorithm* algorithm, Config* config, Params* params)
{
    algorithm->config = config;
    algorithm->params = params;

    for (size_t i = 0; i < NUM_PIXELS; i++) {
        algorithm->force[i] = 1.0;
        algorithm->hueVel[i] = 0.0;
        algorithm->linearSatVel[i] = 0.0;
        algorithm->linearValVel[i] = 0.0;
        algorithm->linearSat[i] = 0.0;
        algorithm->linearVal[i] = 0.0;
        algorithm->hue[i] = sys_rand32_get() / (float)UINT32_MAX;
        algorithm->sat[i] = 0.0;
        algorithm->val[i] = 0.0;
    }
}

void algorithm_update_inner(Algorithm* algorithm, float* position, float* velocity, float target, float tolerance, int wraps)
{
    // Sometimes the fastest way to get to the target is to decrement
    // the hue, and wrap around from 0 -> 1. The "inverse target" is
    // the opposite point on the hue circle. By adding or subtracting
    // 1, we make the comparisons below simpler. Because we wrap the
    // values at the end of this function, the offset here doesn't
    // effect the final value.
    if (wraps) {
        float inverse_target = target + 0.5f;
        if (target > 1.0f) {
            target -= 1.0f;
        }

        if (inverse_target > target) {
            for (size_t i = 0; i < NUM_PIXELS; i++) {
                if (position[i] > inverse_target) {
                    position[i] -= 1.0f;
                }
            }
        } else {
            for (size_t i = 0; i < NUM_PIXELS; i++) {
                if (position[i] < inverse_target) {
                    position[i] += 1.0f;
                }
            }
        }
    }

    // Calculate random forces on the pixel due to temperature
    for (size_t i = 0; i < NUM_PIXELS; i++) {
        algorithm->force[i] = sys_rand32_get() / (float)UINT32_MAX;

        algorithm->force[i] -= algorithm->params->temp;

        if (algorithm->force[i] > 0.0f) {
            algorithm->force[i] = 0.0;
        } else {
            algorithm->force[i] += algorithm->params->temp / 2.0f;
        }

        if (algorithm->force[i] < 0.0f) {
            algorithm->force[i] = -1.0;
        } else if (algorithm->force[i] > 0.0f) {
            algorithm->force[i] = 1.0;
        }

        algorithm->force[i] *= algorithm->config->delta;
    }

    // Apply restoring forces
    float restoring_delta = algorithm->config->delta * algorithm->config->restoringForce;
    if (wraps || target == 0.0f) {
        for (size_t i = 0; i < NUM_PIXELS; i++) {
            if (position[i] > target + tolerance) {
                algorithm->force[i] -= restoring_delta;
            }
        }
    }

    for (size_t i = 0; i < NUM_PIXELS; i++) {
        if (position[i] < target - tolerance) {
            algorithm->force[i] += restoring_delta;
        }
    }

    // Acceleration = Force / Mass
    for (size_t i = 0; i < NUM_PIXELS; i++) {
        algorithm->force[i] /= algorithm->config->mass;
    }

    // Velocity += Acceleration
    for (size_t i = 0; i < NUM_PIXELS; i++) {
        velocity[i] += algorithm->force[i];
    }

    // Position += Velocity
    for (size_t i = 0; i < NUM_PIXELS; i++) {
        position[i] += velocity[i];
    }

    // Velocity *= friction
    for (size_t i = 0; i < NUM_PIXELS; i++) {
        velocity[i] *= algorithm->config->friction;
    }

    if (wraps) {
        for (size_t i = 0; i < NUM_PIXELS; i++) {
            if (position[i] < 0.0f) {
                position[i] += 1.0f;
            } else if (position[i] > 1.0f) {
                position[i] -= 1.0f;
            }
        }
    } else {
        for (size_t i = 0; i < NUM_PIXELS; i++) {
            if (position[i] < 0.0f) {
                position[i] = 0.0f;
            } else if (position[i] > 1.0f) {
                position[i] = 1.0f;
            }
        }
    }
}

void algorithm_update(Algorithm* algorithm)
{
    algorithm_update_inner(algorithm, algorithm->hue, algorithm->hueVel, algorithm->params->hue, algorithm->params->dh, 1);
    algorithm_update_inner(algorithm, algorithm->linearSat, algorithm->linearSatVel, algorithm->params->sat, algorithm->params->ds, 0);
    algorithm_update_inner(algorithm, algorithm->linearVal, algorithm->linearValVel, algorithm->params->val, algorithm->params->dv, 0);

    // This is where we raise it to the power of 1 / 1.2, to bend it above y=x
    // But that's expensive, so I just wont do it.
    for (size_t i = 0; i < NUM_PIXELS; i++) {
        algorithm->sat[i] = algorithm->linearSat[i];
        algorithm->val[i] = algorithm->linearVal[i];
    }
}
